cfy_manager/components/postgresql_server/postgresql_server.py

```diff
@@ -377,6 +386,257 @@ def _configure_cluster(self):
+            node['etcd_state'] = etcd_status['state']
+        else:
+            node['etcd_state'] = 'dead'
+            node['errors'].append('Could not retrieve etcd status')
+
+        return node
+
+    def get_cluster_status(self):
+        # Start healthy, and let the status get worse as we find problems
+        status, exit_code = self._update_db_cluster_status(None, 'healthy')
+
+        master_address, replica_addresses = self._get_cluster_addresses()
+
+        master_status = self._get_node_status(master_address, 'DB')
+        master_etcd_state = self._get_node_status(master_address, 'etcd')
+        master = self._process_node_status(
+            master_address, master_status, master_etcd_state, master=True,
+        )
+
+        sync_ips = []
+        master_replication = master_status.get('replication', [])
+        if master_replication:
+            for replica in master_replication:
+                if replica['sync_state'] == 'sync':
+                    sync_ips.append(replica['client_addr'])
+
+        replicas = []
+        for address in replica_addresses:
+            node_status = self._get_node_status(address, 'DB')
+            etcd_state = self._get_node_status(address, 'etcd')
+            replicas.append(
+                self._process_node_status(
+                    address, node_status, etcd_state,
+                    sync_replica=address in sync_ips,
+                )
+            )
+        db_nodes = [master] + replicas
+
+        etcd_followers = 0
+        etcd_leaders = 0
+        for node in db_nodes:
+            if node['etcd_state'] == 'StateFollower':
+                etcd_followers += 1
+            elif node['etcd_state'] == 'StateLeader':
+                etcd_leaders += 1
+
+        if etcd_leaders != 1:
+            logger.error(
+                'Expected to find 1 etcd leader, but found {num}, '
+                'cluster consensus lost.'.format(
+                    num=etcd_leaders,
+                )
+            )
+            status, exit_code = self._update_db_cluster_status(status, 'down')
+
+        if etcd_followers < 1:
+            logger.error(
+                'No etcd followers found, cluster consensus lost.'
+            )
+            status, exit_code = self._update_db_cluster_status(status, 'down')
+        elif etcd_followers < 2:
+            logger.warning(
+                'Missing one etcd follower.'
+            )
+            status, exit_code = self._update_db_cluster_status(status,
+                                                               'degraded')
+
+        if not master['alive']:
+            status, exit_code = self._update_db_cluster_status(status, 'down')
+
+        if not sync_ips:
+            # The cluster is down if there are no sync replicas, because
+            # writes will not be allowed
+            status, exit_code = self._update_db_cluster_status(status, 'down')
+
+        master_log_location = master.get('log_location')
```

Comment:
as above - can this really not be here? what if it's not?

Reply:
It can be not here, e.g. if a server is just coming up. Nothing much can be done then though.